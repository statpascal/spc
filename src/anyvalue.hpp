/** \file anyvalue.hpp
*/

#pragma once

#include <algorithm>
#include <iostream>

namespace statpascal {

class TAnyValue {
public:
    TAnyValue ();
    TAnyValue (const TAnyValue &);
    TAnyValue (TAnyValue &&);
    TAnyValue (void *p);
    template<typename T, 
             typename U = typename std::remove_cv<typename std::remove_reference<T>::type>::type,
             typename V = typename std::enable_if<!std::is_same<TAnyValue, U>::value>::type> TAnyValue (T &&);
             
    TAnyValue &operator = (TAnyValue);
    
    ~TAnyValue ();
    
    TAnyValue &swap (TAnyValue &);
    
    /** creates copy if value is shared. */
    void copyOnWrite ();
    
    /** returns true if a value has been stored. */
    bool hasValue () const;
    
    template<typename T> T &get ();
    template<typename T> const T &get () const;
    
private:
    class TValue {
    public:
        TValue (): refcount (1) {};
        virtual ~TValue () = default;
        virtual TValue *clone () = 0;
        
        std::size_t refcount;
    };
    
    template<typename T> class TConcreteValue: public TValue {
    public:
        TConcreteValue (const T &);
        TConcreteValue (T &&);
        virtual ~TConcreteValue () = default;
        
        virtual TValue *clone () override;

        T concreteValue;
    };
    
    TValue *value;
};

template<typename T> inline TAnyValue::TConcreteValue<T>::TConcreteValue (const T &val): concreteValue (val) {
}

template<typename T> inline TAnyValue::TConcreteValue<T>::TConcreteValue (T &&val): concreteValue (std::move (val)) {
}

template<typename T> inline TAnyValue::TValue *TAnyValue::TConcreteValue<T>::clone () {
    return new TConcreteValue (concreteValue);
}

inline TAnyValue::TAnyValue (): 
  value (nullptr) {
}

inline TAnyValue::TAnyValue (void *p):
  value (static_cast<TValue *> (p)) {
      if (value)
          ++value->refcount;
}

inline TAnyValue::~TAnyValue () {
    if (value && !--value->refcount)
        delete value;
}

inline TAnyValue::TAnyValue (const TAnyValue &other):
  value (other.value) {
    if (value)
      ++value->refcount;

}

inline TAnyValue::TAnyValue (TAnyValue &&other):
  value (other.value) {
    other.value = nullptr;
}

template<typename T, typename U, typename V> inline TAnyValue::TAnyValue (T &&val):
  value (new TConcreteValue<U> (std::forward<T> (val))) {
}

inline TAnyValue &TAnyValue::operator = (TAnyValue other) {
    return swap (other);
}

inline TAnyValue &TAnyValue::swap (TAnyValue &other) {
    std::swap (value, other.value);
    return *this;
}

template<typename T> inline T &TAnyValue::get () {
    return static_cast<TConcreteValue<T> *> (value)->concreteValue;
}

template<typename T> inline const T &TAnyValue::get () const {
    return static_cast<const TConcreteValue<T> *> (value)->concreteValue;
}

inline void TAnyValue::copyOnWrite () {
    if (value && value->refcount > 1) {
        --value->refcount;
        value = value->clone ();
    }
}

inline bool TAnyValue::hasValue () const {
    return !!value;
}

}

#ifdef TEST

// Test

#include <iostream>

using namespace statpascal;

class TTestClass {
public:
    TTestClass (int n): n (n) { std::cout << "Konstruktor: " << n << std::endl; }
    ~TTestClass () { std::cout << "Destruktor: " << n << std::endl; }
    TTestClass (const TTestClass &other): n (other.n) { std::cout << "Copy Constructor: " << n << std::endl; }
    TTestClass (TTestClass &&other): n (other.n)  { std::cout << "Move Contructor: " << n << std::endl; other.n = 0; }
    
    int getn () { return n;}
private:
    int n;
};

int main () {
//    TAnyValue any (TTestClass (5));
    
    TTestClass t (7);
    TAnyValue any (TTestClass (6));
    std::cout << "Size is: " << sizeof (any) << std::endl << " " << t.getn () << std::endl;
    TAnyValue any1 (std::move (any));
    TAnyValue any2 = any1;
    TAnyValue any3 = any1;
    any1 = any1;
    std::cout << "Any is: " << any1.get<TTestClass> ().getn () << std::endl;
}

#endif
