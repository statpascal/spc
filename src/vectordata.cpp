#include "vectordata.hpp"
#include "anymanager.hpp"

namespace statpascal {

TVectorDataPtr &TVectorDataPtr::operator = (TVectorDataPtr other) {
    return swap (other);
}

TVectorDataPtr &TVectorDataPtr::swap (TVectorDataPtr &other) {
    std::swap (vectorData, other.vectorData);
    return *this;
}

void TVectorDataPtr::copyOnWrite () {
    if (vectorData && vectorData->refcount > 1) {
        --vectorData->refcount;
        TVectorData *copy = static_cast<TVectorData *> (malloc (sizeof (TVectorData) + vectorData->size * vectorData->count));
        new (copy) TVectorData (*vectorData);
        vectorData = copy;
    }
}

TVectorData::TVectorData (const TVectorData &other):
  size (other.size), count (other.count), refcount (1), anyManager (other.anyManager) {
    if (anyManager)
        for (std::size_t i = 0; i < count; ++i)
            setElement (i, &other.get<unsigned char> (i));
    else
        memcpy (data, other.data, size * count);
}

void TVectorData::setElement (std::size_t index, const void *src) {
    unsigned char *dest = &data [index * size];
    std::memcpy (dest, src, size);
    if (anyManager)
        anyManager->copy (src, dest);
}

void TVectorData::deleteData () {
    if (anyManager)        
        for (std::size_t i = 0; i < count; ++i)
            anyManager->destroy (&get<unsigned char> (i));
}

}