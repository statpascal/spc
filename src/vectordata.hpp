/** \file vectordata.hpp
*/

#pragma once

#include <cstddef>
#include <cstring>
#include <cstdlib>
#include <new>
#include <algorithm>

namespace statpascal {

class TVectorData;
class TAnyManager;

class TVectorDataPtr {
public:
    TVectorDataPtr ();
    TVectorDataPtr (const TVectorDataPtr &);
    explicit TVectorDataPtr (void *, bool incRef = true);
    TVectorDataPtr (std::size_t elementSize, std::size_t elementCount, TAnyManager *elementAnyManager = nullptr, bool zeroMemory = true);
    
    TVectorDataPtr &operator = (TVectorDataPtr);
    
    ~TVectorDataPtr ();

    TVectorDataPtr &swap (TVectorDataPtr &);
    
    void copyOnWrite ();
    bool hasValue () const;
    
    TVectorData &operator * () const;
    TVectorData *operator -> () const;
    
    TVectorData *makeRef () const;
    void releaseRef ();
    
private:
    TVectorData *vectorData;
};


class TVectorData {
public:
    void setElement (std::size_t index, const void *src);
    const void *getElement (std::size_t index) const;
    void *getElement (std::size_t index);
    
    template<typename T> T &get (std::size_t index);
    template<typename T> const T &get (std::size_t index) const;
    
    TAnyManager *getElementAnyManager () const;
    std::size_t getElementSize () const;
    std::size_t getElementCount () const;
    std::size_t getRefCount () const;

private:
    TVectorData (std::size_t elementSize, std::size_t elementCount, TAnyManager *elementAnyManager = nullptr, bool zeroMemory = true);
    TVectorData (const TVectorData &);
    ~TVectorData ();
    
    TVectorData &operator = (TVectorData) = delete;
    
    void deleteData ();

    std::size_t size, count, refcount;
    TAnyManager *anyManager;
    unsigned char data [1];
    
    friend TVectorDataPtr;
};


inline TVectorData::TVectorData (std::size_t size, std::size_t count, TAnyManager *anyManager, bool zeroMemory):
  size (size), count (count), refcount (1), anyManager (anyManager) {
    if (zeroMemory)
        std::fill (data, data + size * count, 0);
}

inline TVectorDataPtr::TVectorDataPtr (std::size_t elementSize, std::size_t elementCount, TAnyManager *elementAnyManager, bool zeroMemory):
  vectorData (static_cast<TVectorData *> (malloc (sizeof (TVectorData) + elementSize * elementCount))) {
    new (vectorData) TVectorData (elementSize, elementCount, elementAnyManager, zeroMemory);
}

inline TVectorData *TVectorDataPtr::operator -> () const {
    return vectorData;
}

inline TVectorData *TVectorDataPtr::makeRef () const {
    if (vectorData)
        ++vectorData->refcount;
    return vectorData;
}

inline void TVectorDataPtr::releaseRef () {
    if (vectorData && !--vectorData->refcount) {
        vectorData->~TVectorData ();
        free (vectorData);
    }
}

inline TVectorData::~TVectorData () {
    if (anyManager)
        deleteData ();
}

inline TAnyManager *TVectorData::getElementAnyManager () const {
    return anyManager;
}

inline std::size_t TVectorData::getElementSize () const {
    return size;
}

inline std::size_t TVectorData::getRefCount () const {
    return refcount;
}

inline std::size_t TVectorData::getElementCount () const {
    return count;
}

inline const void *TVectorData::getElement (std::size_t index) const {
    return &data [index * size];
}

inline void *TVectorData::getElement (std::size_t index) {
    return &data [index * size];
}

template<typename T> inline const T &TVectorData::get (std::size_t index) const {
    return *static_cast<const T *> (getElement (index));
}

template<typename T> inline T &TVectorData::get (std::size_t index) {
    return const_cast<T &> (static_cast<const TVectorData *> (this)->get<T> (index));
}

inline TVectorDataPtr::TVectorDataPtr ():
  vectorData (nullptr) {
}

inline TVectorDataPtr::TVectorDataPtr (const TVectorDataPtr &other):
  vectorData (other.makeRef ()) {
}

inline TVectorDataPtr::TVectorDataPtr (void *p, bool incRef):
  vectorData (static_cast<TVectorData *> (p)) {
    if (incRef && vectorData)
        ++vectorData->refcount;
}

inline TVectorDataPtr::~TVectorDataPtr () {
    releaseRef ();
}

inline bool TVectorDataPtr::hasValue () const {
    return !!vectorData;
}

inline TVectorData &TVectorDataPtr::operator * () const {
    return *vectorData;
}

}
