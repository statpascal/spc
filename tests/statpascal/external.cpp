// gcc -fPIC -I../../src -shared -o external.so external.cpp

#include <stdio.h>
#include <string>
#include <iostream>
#include <cmath>

#include "anyvalue.hpp"
#include "vectordata.hpp"

extern "C" { 

int add (int a, int b) {
    return a + b;
}

int mul (int a, int b) {
    return a * b;
}

void inc (int *a) {
    ++*a;
}

struct rec {
    int a;
    long long b;
    char c;
    double d;
    bool e;
    struct {
        unsigned char g;
        unsigned short h;
    } f;
};

void printrec (rec *r) {
    printf ("%d %lld %c %g %d %d %d\n", r->a, r->b, r->c, r->d, r->e, r->f.g, r->f.h);
}

void changerec (rec *r) {
    r->a = 7;
    r->b = -1;
    r->c = 'D';
    r->d = 2.71828;
    r->e = true;
    r->f.h = 99;
}

void printstring (const char *p) {
    puts (p);
}

void printstring1 (statpascal::TAnyValue p) {
    if (p.hasValue ())
        std::cout << p.get<std::string> ();
}

void setstring (statpascal::TAnyValue *p) {
    std::string s = "A string defined in C++";
    *p = statpascal::TAnyValue (std::move (s));
}

void modifystring (statpascal::TAnyValue *p) {
    if (p->hasValue ()) {
        p->copyOnWrite ();
        for (std::string::value_type &c: p->get<std::string> ())
            c = std::toupper (c);
    }
}

}
