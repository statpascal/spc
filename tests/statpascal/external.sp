program extern;

const
    libname = 'external.so';

type 
    rec = record
        a: integer;
        b: int64;
        c: char;
        d: real;
        e: boolean;
        f: record
            g: byte;
            h: word
        end 
    end; 
    intptr = ^integer;
    charptr = ^char;

function add (a, b: integer): integer; external libname;
function multiply (a, b: integer): integer; external libname name 'mul';

procedure inc (var a: integer); external libname;
procedure incptr (b: intptr); external libname name 'inc';

procedure printrec (var r: rec); external libname;
procedure changerec (var r: rec); external libname;

procedure printstring (p: charptr); external libname;
procedure printstring1 (s: string); external libname;

procedure setstring (var s: string); external libname;
procedure modifystring (var s: string); external libname;

function sub (a, b: integer): integer;
    begin
        sub := a - b
    end;

var 
    a: integer;
    r: rec;
    s: string;
    i: int64;

begin
    writeln (add (2, 3));
    writeln (multiply (6, 7));
    writeln (sub (2, 3));
    a := 5;
    inc (a);
    writeln (a);
    incptr (addr (a));
    writeln (a);
    r.a := 1;
    r.b := -5;
    r.c := 'A';
    r.d := 4.0 * arctan (1.0);
    r.e := false;
    r.f.g := 31;
    r.f.h := 71;
    printrec (r);
    changerec (r);
    writeln (r.a, ' ', r.b, ' ', r.c, ' ', r.d, ' ', r.e, ' ', r.f.g, ' ', r.f.h);
    s := 'Hello world';
    printstring (addr (s [1]));
    printstring1 (s);
    writeln;
    setstring (s);
    writeln (s);
    modifystring (s);
    writeln (s)
end.
