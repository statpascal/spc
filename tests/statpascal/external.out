5
42
-1
6
7
1 -5 A 3.14159 0 31 71
7 -1 D 2.71828 TRUE 31 99
Hello world
Hello world
A string defined in C++
A STRING DEFINED IN C++
